File: BigData/extraction_scripts/mongodb/src/main/java/org/tonde/Immatriculation.java
package org.tonde;

import com.mongodb.client.MongoCollection;
import com.mongodb.client.MongoDatabase;
import com.mongodb.client.model.CreateCollectionOptions;
import com.mongodb.client.model.Filters;
import com.mongodb.client.model.ValidationOptions;
import org.bson.BsonType;
import org.bson.Document;
import org.bson.conversions.Bson;

import java.util.ArrayList;

public class Immatriculation {

    public static void createCollImmatriculation(MongoDatabase db) {

        Bson immatriculationType = Filters.type("Immatriculation", BsonType.STRING);
        Bson marqueType = Filters.type("Marque", BsonType.STRING);
        Bson nomType = Filters.type("Nom", BsonType.STRING);
        Bson puissanceType = Filters.type("Puissance", BsonType.DOUBLE);
        Bson longueurType = Filters.type("Longueur", BsonType.STRING);
        Bson nbPlacesType = Filters.type("NbPlaces", BsonType.INT32);
        Bson nbPortesType = Filters.type("NbPortes", BsonType.INT32);
        Bson Couleurtype = Filters.type("Couleur", BsonType.STRING);
        Bson Occasiontype = Filters.type("Occasion", BsonType.BOOLEAN);
        Bson Prixtype = Filters.type("Prix", BsonType.DOUBLE);


        Bson validator = Filters.and(immatriculationType, marqueType, nomType, puissanceType,longueurType,nbPlacesType,nbPortesType,Couleurtype,Occasiontype,Prixtype);

        db.getCollection("Immatriculation").drop();
        db.createCollection("Immatriculation", new CreateCollectionOptions().validationOptions(new ValidationOptions()
                .validator(validator)));
    }

    public static void insertImmatriculation (ArrayList<String[]> immatriculationArr, MongoCollection<Document> collection) {
        for (String[] vehicule : immatriculationArr) {
            Document doc = new Document()
                    .append("Immatriculation",vehicule[0])
                    .append("Marque",vehicule[1])
                    .append("Nom",vehicule[2])
                    .append("Puissance",Float.parseFloat(vehicule[3]))
                    .append("Longueur",vehicule[4])
                    .append("NbPlaces",Integer.parseInt(vehicule[5]))
                    .append("NbPortes",Integer.parseInt(vehicule[6]))
                    .append("Couleur",vehicule[7])
                    .append("Occasion",Boolean.parseBoolean(vehicule[8]))
                    .append("Prix",Float.parseFloat(vehicule[9]));

            collection.insertOne(doc);
        }
    }

}

File: BigData/extraction_scripts/insertion_catalogue_immatriculation/immatriculation.py
from pymongo import MongoClient
from pymongo.database import Database
import csv

class Immatriculation:

    @staticmethod
    def create_coll_immatriculation(db: Database):
        validator = {
            "$jsonSchema": {
                "bsonType": "object",
                "required": ["Immatriculation", "Marque", "Nom", "Puissance", "Longueur", "NbPlaces", "NbPortes", "Couleur", "Occasion", "Prix"],
                "properties": {
                    "Immatriculation": {"bsonType": "string"},
                    "Marque": {"bsonType": "string"},
                    "Nom": {"bsonType": "string"},
                    "Puissance": {"bsonType": "double"},
                    "Longueur": {"bsonType": "string"},
                    "NbPlaces": {"bsonType": "int"},
                    "NbPortes": {"bsonType": "int"},
                    "Couleur": {"bsonType": "string"},
                    "Occasion": {"bsonType": "bool"},
                    "Prix": {"bsonType": "double"}
                }
            }
        }

        db.drop_collection("Immatriculation")
        db.create_collection("Immatriculation", validator=validator)

    @staticmethod
    def insert_immatriculation(immatriculation_arr, collection):
        for vehicule in immatriculation_arr:
            doc = {
                "Immatriculation": vehicule[0],
                "Marque": vehicule[1],
                "Nom": vehicule[2],
                "Puissance": float(vehicule[3]),
                "Longueur": vehicule[4],
                "NbPlaces": int(vehicule[5]),
                "NbPortes": int(vehicule[6]),
                "Couleur": vehicule[7],
                "Occasion": vehicule[8].lower() == 'true',
                "Prix": float(vehicule[9])
            }
            collection.insert_one(doc)

def load_immatriculation_data_from_file(file_path):
    immatriculation_arr = []
    with open(file_path, mode='r') as file:
        reader = csv.reader(file)
        next(reader)  # Skip header row
        for row in reader:
            immatriculation_arr.append(row)
    return immatriculation_arr

if __name__ == "__main__":
    client = MongoClient("mongodb://localhost:27017/")
    db = client["bigdata_exam"]

    Immatriculation.create_coll_immatriculation(db)
    immatriculation_data = load_immatriculation_data_from_file("Immatriculation.csv")
    collection = db["Immatriculation"]
    Immatriculation.insert_immatriculation(immatriculation_data, collection)
